### DISCUSSION TITLE
Procesamiento masivo desde un mismo cliente

### DISCUSSION THREAD

**Discussion by jarufe** [manudona92/MapantES#11] [Ideas]
Denomino procesamiento masivo a dos características que se pueden dar a la vez o por separado: tener más de 1 hilo de procesamiento en el mismo ordenador teniendo lugar al mismo tiempo; disponer de un volumen de datos muy grande "más cerca" para reducir carga por tener que bajarlos desde otro lugar.
Estoy pensando, fundamentalmente, en la posibilidad de tener un ordenador de grandes prestaciones que fuera capaz de tener un número muy alto de hilos de procesamiento a la vez. Por ejemplo, se podría preparar los LAZ y vectoriales de toda una Comunidad Autónoma en un disco y ofrecérselo a ese ordenador. El cliente lanzaría n hilos (según la capacidad del ordenador y también del servidor de atender a peticiones de los datos que le hacen falta a cada hilo). Teniendo los datos en local, el servidor le ofrecería una lista de los ficheros a cada hilo y el hilo se encargaría de obtenerlos en local.
Por lo tanto, conceptualmente habría 3 servidores: el servidor que ante una petición asigna una cuadrícula y le dice al hilo del  cliente qué ficheros le hacen falta (el que ya hay ahora); el servidor de almacenamiento donde están alojados los ficheros a procesar (se busca que el IGN nos dé un método por URL, pero para este caso podría ser una ubicación local, o en red); el servidor de almacenamiento donde irían a guardarse los ficheros resultado del procesamiento
Podría haber un cuarto servidor de almacenamiento de parámetros de procesamiento, pero normalmente sería el mismo que el primero.
Así es como concebí mi solución en Java y me parece que podría ser muy eficiente para trabajar con una cantidad enorme de datos como es el proyecto de toda España.

**Comment by jarufe** [comment]
Dejo aquí los datos de la prueba que hice con la provincia de A Coruña según mi concepción de lo que podría ser un procesamiento masivo:

- Equipamiento: 4 PCs en red, 4 hilos cada uno = 16 hilos de procesamiento
    Intel i5-4440, 3.10GHz, 4GB RAM
    Conectados en red a través de Switch 3COM. 
    Switch conectado a Internet a través de Router Movistar de fibra óptica
- Cuadrículas y vectoriales en local
    Primer intento con los ficheros en uno de los PCs. Al no ser un verdadero servidor de almacenamiento, el copiado de múltiples ficheros a los hilos de los 4 PCs no funcionaba bien
    Segundo intento (definitivo) replicando todos los ficheros en cada PC. Cada hilo hacía un copiado simple de ficheros desde una ubicación de su propio PC
- Consulta a servidor en Internet para asignación de cuadrícula de forma aleatoria
    Devuelve un listado de los ficheros necesarios (LAZ principal, LAZ vecinos, vectoriales, INI, TXT)
- Resultados del procesamiento (PNG x 4 tipos + ZIP con curvas y piedras en DXF)
    4 tipos PNG: sin depresión vs con depresión / sin vectorial vs con vectorial
- Almacenamiento del resultado en uno de los PCs de la red


Total Provincia = 2260
Cuadrículas Procesadas = 2107
Cuadrículas con Error =  153 (LAZ no existente en local)
  El error es inmediato al no encontrar el fichero. Pérdida de tiempo despreciable

Superficie A Coruña: 7950 Km2
Superficie procesada (aprox.): 7400 Km2
Tiempo (aprox.): 60h

Espacio ocupado:
  Ficheros LAZ: 54 GB (2107 ficheros)
  Ficheros vectoriales: 175 MB (93 ficheros)
  Resultados:
    PNG + PGW: 11,1 GB (el total de los 4 tipos de PNG)
    ZIP: 6,16 GB (2107 ficheros)

**Comment by manudona92** [comment]
Buena Javi. Buenas estadísticas y enfoque. 

Mis preguntas sería, con una futura opción de multi proceso en lo creado por Agustín ¿Tendrías los mismos rendimientos que con el que has hecho previamente?. Al final y al cabo el proceso de descarga lo tendrás que hacer igual , te saltas el buffer eso sí porque estás en modo batch (el buffer no tarda mucho, pero añade), y luego tendrías que tener más espacio en local, porque nos queremos quedar con los DXF y otros tantos PNGs. 

Lo digo porque digamos que la cimentación de la parte geográfica (proto visor, conteado de tiles procsadas/no, etc.) y vectorial (renderización, exportar DXF, PNG's específicos, etc.) ya está bastante avanzada. Si con la opción de multi proceso que saque Agustín, puedes seguir con los 4 PCs + 4 hilos, creo que seria un buen ordenador para tenerlo de back-up y que agilizará el procesamiento mucho. Al final me salen tus numeros bien: 16 x 50MB = 800 MB/h aprox, que es 4-5 veces más eficiente con mi ordenador 1 PC y 6 procesos (estaba en torno a los 180 MB/h).

Algunos números de cuanto tardaría en procesar algunas CCAA con ese rendimiento: 
La Rioja: 12 dias
Pais Vasco: 2.5 dias
Madrid: 9 días
Cataluña: 10 días
Murcia: 3 días

Otra cosa que me dijo Jarkko, que se me ha pasado y solo hice una vez, es indexar los LAZ con lasindex de LAStools cunado procesamos en **modo BATCH** (más info aquí https://rapidlasso.com/2012/12/03/lasindex-spatial-indexing-of-lidar-data/). En Lasindex metes de input tus archivos LAZ  y se te generará un archivo LAX por cada LAZ (lasindex -cpu64 -i ). Nunca he hecho la comparación si va más rápido. 

Otra es hacer los tiles más pequeños, de 2 km a 1 km, lo que reduciria el tamaño medio de tile de 40MB a 10MB, y en teoría tendría que procesar más rápido. 

Deshecho la opción de indexar para lo de Agustín ya que no procesamos en modo batch (no sé si serviria), además de la de reducir el tamaño del tile, ya que tenemos todoso los ID y SHP organizado por el IGN y sería más follón. Los lotes más densos (La Rioja, Madrid) ya de por sí están en 1 x 1 km.

**Reply by jarufe** [reply]
Te respondo a bote pronto (no hice cálculos antes de ponerme a escribir)
Mi primer impulso es decir que voy a seguir teniendo unos rendimientos similares. Pero ojo con la palabra "similar", una variación de unos pocos segundos por cuadrícula acaba siendo de horas en el procesamiento de una provincia como A Coruña considerando un solo hilo de procesamiento.
Una primera variación entre la prueba que hice y lo real con el mecanismo desarrollado por Agustín va a ser que yo no tuve que descargar los ficheros, pues los tenía en local. Aunque mi conexión a internet sea rápida, me imagino que tardaré más tiempo en bajar los ficheros que en copiarlos.
Una segunda variación no es en si porque el mío se haga en modo batch. Cada hilo de procesamiento solo trabaja con la cuadrícula y sus vecinos, al igual que hace Agustín. Por eso me parece que hacer el lasindex no va a tener repercusión en mi forma de trabajar. lasindex tiene sentido, creo yo, si en tu carpeta "in" vas a tener tropecientos ficheros LAZ que se van a ir procesando hasta que se acabe con todos. Para cada cuadrícula vas a tener a sus vecinos pero también a un montón de ficheros que no son vecinos, los cuales KP va a tener que descartar porque no los necesita para hacer el buffer. Para eso es importante tenerlos indexados. Pero si solo tienes los ficheros que ya sabes que necesitas, no sé si el hecho de indexarlos previamente va a dar mucha mejora, más que nada porque el indexado también tardará un tiempo en hacerse.
El espacio necesario no creo que represente una variación. Cada hilo realiza todo el procesamiento y genera todos los productos, los que queremos y los que no. Luego, los que queremos se suben al servidor (en mi prueba se dejaban en uno de los PCs en red). Por último, todo lo que ha procesado el hilo se borra; con el cliente de Agustín también se borra todo al final, ¿no?
En cuanto a hacer los tiles más pequeños, creo que es importante lo que dices del follón que les podemos generar a los del IGN. Se puede buscar cual es la mejor relación tamaño / esfuerzo de procesamiento, pero me parece que podemos asumir las cosas tal y como están, ¿no?
Los PCs que usé para la prueba tal vez podría seguir usándolos, incluso alguno más. No representarían un back-up sino que sería lo que yo podría aportar en lugar de un mísero portátil que se cae a trozos. Cada PC tendría en ejecución el cliente de Agustín como si fueran distintos colaboradores desde sus casas, y ojalá que con multiproceso para sacarles más partido.

**Reply by manudona92** [reply]
Vale, me que claro la parte del indexado. 

Entiendo que con lo de Agustín también se borra todo menos los 3-4 PNG's y los DXF que nos interesan. Ya que en caso de renderizar de nuevo lo necesitariamos de backup, también porque la idea es comprimir los DXF (curva de nivel) y tenerlos el dia de mañana.

Respecto tiles, estoy de acuerdo. Los de mayor densidad, que ahora son Madrid y La RIoja ya estan a 1km, y reduce el peso considerablemente.
